### PR TITLE
Fix some issues with painless's strings

### LINKED ISSUE #22372: Double backslash in Painless script strings causes a syntax error
**Elasticsearch version**: 5.0.2

**Plugins installed**: X-Pack

**JVM version**: 1.8.0_111-b15

**OS version**: CentOS 7.2.1511

**Description of the problem including expected versus actual behavior**:
It seems that using double backslash in Painless script string values causes a syntax error. I need to perform a comparison against values with double backslashes. Is there a workaround for this, am I doing something wrong or is it a bug?

**NOTE**: the original error message below contains double backslashes in '...['\\A]...' and '...{ '\\Archived' }...', but it gets parsed by GitHub into a single backslash.

**Steps to reproduce**:
```
GET myindex/_search
{
  "size": 50,
  "script_fields": {
    "new_field": {
      "script": {
        "lang": "painless",
        "inline": "if (doc['finalstatus.keyword'].value == '\\Archived') { 'Hard Archived' } else { doc['otherfield.keyword'] }"
      }
    }
  }
}
```

**Provide logs (if relevant)**:
```
"failed_shards": [
      {
        "shard": 0,
        "index": "myindex",
        "node": "1LwGYO3RQB6tOYKB9Mgo_A",
        "reason": {
          "type": "script_exception",
          "reason": "compile error",
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "unexpected character ['\\A].",
            "caused_by": {
              "type": "lexer_no_viable_alt_exception",
              "reason": "lexer_no_viable_alt_exception: null"
            }
          },
          "script_stack": [
            "... ectDownloadCompleted') { '\\Archived' } else { doc[ ...",
            "                             ^---- HERE"
          ]
...
```

### PR DESCRIPTION
**Pull request by nik9000**
1. Escape sequences we're working. For example `\\` is now correctly
interpreted as `\` instead of `\\`. Same with `\'` being `'` and
`\"` being `"`.
2. `'` delimited strings weren't allowed to contain `"`s but it looked
like they were intended to support it. Now they do.
3. Improves the error message when the script contains an invalid
escape sequence inside a string to include a list of the valid
escape sequences.

Closes #22372

### CODE REVIEW

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/BasicExpressionTests.java [hunk 1]
```diff
@@ -55,6 +56,33 @@ public void testConstantCharTruncation() {
...
+        // We don't break native escapes like new line
+        assertEquals("\nstring", exec("\"\nstring\""));
+        assertEquals("\nstring", exec("'\nstring'"));
+    }
+
+    public void testStringTermination() {
+        // `'` inside of a string delimited with `"` should be ok
+        assertEquals("test'", exec("\"test'\""));
+        // `"` inside of a string delimited with `'` should be ok
+        assertEquals("test\"", exec("'test\"'"));
```
Comment (nik9000) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/BasicExpressionTests.java]: This one didn't work without the lexer change. I think we intended it to work.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedPainlessLexer.java [hunk 2]
```diff
@@ -77,7 +78,15 @@ public void recover(final LexerNoViableAltException lnvae) {
         final String text = charStream.getText(Interval.of(startIndex, charStream.index()));
 
         Location location = new Location(sourceName, _tokenStartCharIndex);
-        throw location.createError(new IllegalArgumentException("unexpected character [" + getErrorDisplay(text) + "].", lnvae));
+        String message = "unexpected character [" + getErrorDisplay(text) + "].";
+        char firstChar = text.charAt(0);
+        if ((firstChar == '\'' || firstChar == '"') && text.length() - 2 > 0 && text.charAt(text.length() - 2) == '\\') {
```
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedPainlessLexer.java]: I thought about using fancy lexer changes to make this more correct or allowing all escapes in the lexer but detecting invalid escapes in the walker. I'm not sure which is better but I'd already written this one so I figured I start with it.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java [hunk 3]
```diff
@@ -799,7 +799,19 @@ public ANode visitNull(NullContext ctx) {
 
     @Override
     public ANode visitString(StringContext ctx) {
-        String string = ctx.STRING().getText().substring(1, ctx.STRING().getText().length() - 1);
+        String string = ctx.STRING().getText();
+        boolean doubleQuoted = string.charAt(0) == '"';
+
+        // Strip the leading and trailing quotes
+        string = string.substring(1, ctx.STRING().getText().length() - 1);
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: Can we do all of these things (stripping last char, backslash escape and quote escape) in one pass with a string builder? Right 3 copies are made of the string.

Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: We can. I figured the replacements would mostly end up as noops so it wasn't worth it. Especially with this on the compile side. If you'd prefer it I'm happy to do it though.

**Comment by nik9000** [comment]
@rjernst, I replaced the replaces with walking a string builder. I expect I could save some CPU by copying the strings in longer chunks if possible but I figure this performance is adequate and more readable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java [hunk 4]
```diff
@@ -799,9 +799,25 @@ public ANode visitNull(NullContext ctx) {
 
     @Override
     public ANode visitString(StringContext ctx) {
-        String string = ctx.STRING().getText().substring(1, ctx.STRING().getText().length() - 1);
+        StringBuilder string = new StringBuilder(ctx.STRING().getText());
+
+        // Strip the leading and trailing quotes and replace the escape sequences with their literal equivalents
+        int src = 1;
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: Add asserts for the leading and trailing quotes?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java (hunk 4 above)
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java]: Can do.

**Comment by nik9000** [comment]
Thanks for reviewing @rjernst!

master: f24ca5188a39a9d8bd48f76bb7cbae2f7897e83d
5.x: 55aed742471e8b27b88cbdac22f017c16109240e

**Comment by clintongormley** [comment]
@nik9000 shouldn't this be backported to 5.2?

**Comment by nik9000** [comment]
@clintongormley, probably. I'm not sure why I didn't do it automatically. The workaround I mention in 
#22372 works but it isn't great. I'll backport.

**Comment by nik9000** [comment]
5.2: 3a80f7421d9eabbb09a89c087c5de0387c403e7b

**Pull request merged**